### PR TITLE
ISD-2142 Render default haproxy config

### PR DESCRIPTION
**Pull request by Thanhphan1147**
Applicable spec: ISD-163(Internal)

Adds implementation for the `config_changed` hook to render the templated `haproxy.cfg` file. This PR also add rendering of the config file for Diffie-Hellman key exchange.

Update unit test.

The supported workflow for now is:
1. Deploying the haproxy-operator charm
2. When connecting to one of the haproxy unit's address, a default response will be returned

```sh
juju deploy ./haproxy_amd64.charm

curl <juju_unit_ip>:80
# we should get Default page for the haproxy-operator charm
```

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Comment by Thanhphan1147** [comment]
@gregory-schiano can you take a look at this PR and leave your review? Thanks!

File: src/haproxy.py [hunk 1]
```diff
@@ -46,3 +73,25 @@ def is_active(self) -> bool:
...
+        """
+        path.write_text(content, encoding="utf-8")
+        os.chmod(path, mode)
+        u = pwd.getpwnam(HAPROXY_USER)
+        # Set the correct ownership for the file.
+        os.chown(path, uid=u.pw_uid, gid=u.pw_gid)
+
+    def _render_haproxy_config(self) -> None:
+        """Render the haproxy configuration file."""
+        with open("templates/haproxy.cfg.j2", "r", encoding="utf-8") as file:
```
Comment (yanksyoon) [inline src/haproxy.py:97]: I think we can use Jinja's loaders :) https://jinja.palletsprojects.com/en/3.0.x/api/?highlight=filesystemloader#jinja2.FileSystemLoader

Comment (yanksyoon) [reply src/haproxy.py:97]: Also it seems that this file does not contain any template replacement, could this be just a cfg file?

**Review by yanksyoon — DISMISSED** [review]
Minor comments, LGTM!

File: src/haproxy.py (hunk 1 above)
Comment (Thanhphan1147) [reply src/haproxy.py:97]: It does look a bit weird to have a jinja2 without any replacements but I'm adding them in the following PRs :smile: so I think the file can be left as-is for now.

**Review by weiiwang01 — DISMISSED** [review]

**Review by weiiwang01 — DISMISSED** [review]

File: src/haproxy.py [hunk 2]
```diff
@@ -4,12 +4,32 @@
...
+from jinja2 import Template
 
 APT_PACKAGE_VERSION = "2.8.5-1ubuntu3"
 APT_PACKAGE_NAME = "haproxy"
+HAPROXY_CONFIG_DIR = Path("/etc/haproxy")
+HAPROXY_CONFIG = Path(HAPROXY_CONFIG_DIR / "haproxy.cfg")
+HAPROXY_USER = "haproxy"
+# Configuration used to parameterize Diffie-Hellman key exchange.
+# Source: https://ssl-config.mozilla.org/ffdhe2048.txt.
+HAPROXY_DH_PARAM = (
```
Comment (amandahla) [inline src/haproxy.py:25]: Could you add more info on how was this generated and why is hard-coded please?

Comment (Thanhphan1147) [reply src/haproxy.py:25]: Yes, I added a more detailed explanation for this.

**Review by weiiwang01 — APPROVED** [review]

**Review by yanksyoon — APPROVED** [review]
LGTM, it looks like the current integration tests should cover the changes? :)

**Pull request merged**